tests/StringHelperTest.php

```diff
@@ -512,4 +512,131 @@ public function testParsePathWithDelimiterEqualsEscapeCharacter(): void
 
         StringHelper::parsePath('key1.key2.key3', '.', '.');
     }
+
+    /**
+     * @return array [$string, $expected]
+     */
+    public function dataTrim(): array
+    {
+        $bom = "\u{FEFF}"; // "\xEF\xBB\xBF"
+        $nbsp = "\u{00A0}"; // "\xC2\xA0"
+        $emsp = "\u{2003}"; // "\xE2\x80\x83"
+        $thsp = "\u{2009}"; // "\xE2\x80\x89"
+        $lsep = "\u{2028}"; // "\xE2\x80\xA8"
+        $ascii = " \f\n\r\t\v\x00";
+
+        $base = 'Здесь我' . $nbsp . '-' . $thsp . 'Multibyte我' . $lsep . 'Строка 👍🏻';
+
+        return [
```

Comment:
Please use `yield` on each item instead

Reply:
I'd also add `dataNoTrim` with the following cases:
```
[$ascii . $ascii . $nbsp . $emsp . $emsp . PHP_EOL, ''],
[$base, $base],
```
Instead of duplicating them in each data provider. 
Also L- and R- `trim` data providers are cases for just `trim` function. So they can be reused.

So then your tests will look like:
```php
     /**
      * @dataProvider dataNoTrim
      * @dataProvider dataLTrim
      * @dataProvider dataRTrim
      * @dataProvider dataTrim
      */
     public function testTrim(string $string, string $expected): void
```

Reply:
> Also L- and R- `trim` data providers are cases for just `trim` function.
no, `ltrim` may skip whitespaces to the right.

Reply:
That's rights. 
Possible cases:
1. What can trim left side: `ltrim` / `trim`
2. What can trim right side: `rtrim` / `trim`
3. What can trim both sides: `trim`

So potentially `trim` can use all `ltrim` and `rtrim` test cases, isn't it?

Reply:
Test cases for `ltrim()`, `rtrim()` have strings with whitespaces on the other side.

`ltrim('  '.$str."\n") = '$str."\n";`
!=
`trim('  '.$str."\n") = $str;`

Reply:
Got it thanks